File: java/kolone.java
import java.util.Scanner;

public class kolone {
	static String getReverseString(String str) {
		return new StringBuilder(str).reverse().toString();
	}

	public static void main(String[] args) {
		Scanner scan = new Scanner(System.in);
		int ant1_length = scan.nextInt();
		int ant2_length = scan.nextInt();
		String ant1 = scan.next();
		String ant2 = scan.next();
		int pos = scan.nextInt();
		if (pos >= ant1_length + ant2_length)
			System.out.println(ant2 + getReverseString(ant1));
		else {
			StringBuilder sb = new StringBuilder(getReverseString(ant1) + ant2);
			for (int i = 0; i < ant1.length() && i < pos; i++) {
				int index = Math.min(sb.length() - i - 1, ant1.length() + pos - 2 * i - 1);
				sb.setCharAt(index, ant1.charAt(i));
			}
			for (int i = 0; i < ant2.length() && i < pos; i++) {
				int index = Math.max(i, ant1.length() - pos + 2 * i);
				sb.setCharAt(index, ant2.charAt(i));
			}
			System.out.println(sb);
		}
		scan.close();
	}
}

File: ts/src/kolone.ts
import readline = require('readline')

function reverseString (str: string): string {
  return str.split('').reverse().join('')
}

function processAnts (ant1: string, ant2: string, seconds: number): string {
  if (seconds >= ant1.length + ant2.length - 1) {
    return `${ant2}${reverseString(ant1)}`
  }

  const str: string[] = [...reverseString(ant1)].concat(...ant2)
  for (let i = 0; i < Math.min(ant1.length, seconds); i++) {
    const index = Math.min(str.length - i - 1, ant1.length + seconds - 2 * i - 1)
    str[index] = ant1.charAt(i)
  }
  for (let i = 0; i < Math.min(ant2.length, seconds); i++) {
    const index = Math.max(i, ant1.length - seconds + 2 * i)
    str[index] = ant2.charAt(i)
  }

  return str.join('')
}

let lineCount = 0
let ant1: string
let ant2: string
let seconds: number

readline
  .createInterface({
    input: process.stdin,
    output: process.stdout
  })
  .on('line', (line: string) => {
    if (lineCount === 1) {
      ant1 = line
    } else if (lineCount === 2) {
      ant2 = line
    } else if (lineCount === 3) {
      seconds = parseInt(line)
    }
    lineCount++
  })
  .on('close', () => {
    console.log(processAnts(ant1, ant2, seconds))
    process.exit(0)
  })
